r/tests/testthat/test-dplyr-funcs-datetime.R

```diff
@@ -1471,3 +1471,128 @@ test_that("make_difftime()", {
     )
   )
 })
+
+test_that("parse_date_time()", {
+  test_df <- tibble(
+    dates = c("09-01-17", "02-Sep-17")
+  )
+
+  test_dates <- tibble::tibble(
+    string_ymd = c(
+      "2021-09-1", "2021/09///2", "2021.09.03", "2021,09,4", "2021:09::5",
+      "2021 09   6", "21-09-07", "21/09/08", "21.09.9", "21,09,10", "21:09:11",
+      # not yet working for strings with no separators, like "20210917", "210918" or "2021Sep19
+      # no separators and %b or %B are even more complicated (and they work in
+      # lubridate). not to mention locale
+      NA
+    ),
+    string_dmy = c(
+      "1-09-2021", "2/09//2021", "03.09.2021", "04,09,2021", "5:::09:2021",
+      "6  09  2021", "07-09-21", "08/09/21", "9.09.21", "10,09,21", "11:09:21",
+      # not yet working for strings with no separators, like "10092021", "100921",
+      NA
+    ),
+    string_mdy = c(
+      "09-01-2021", "09/2/2021", "09.3.2021", "09,04,2021", "09:05:2021",
+      "09 6 2021", "09-7-21", "09/08/21", "09.9.21", "09,10,21", "09:11:21",
+      # not yet working for strings with no separators, like "09102021", "091021",
+      NA
+    )
+  )
+
+  test_dates_locale <- tibble::tibble(
+    string_ymd = c(
+      "2021 Sep 12", "2021 September 13", "21 Sep 14", "21 September 15", NA
+    ),
+    string_dmy = c(
+      "12 Sep 2021", "13 September 2021", "14 Sep 21", "15 September 21", NA
+    ),
+    string_mdy = c(
+      "Sep 12 2021", "September 13 2021", "Sep 14 21", "September 15 21", NA
+    )
+  )
+
+  skip_if_not_available("re2")
+  compare_dplyr_binding(
+    .input %>%
+      mutate(
+        parsed_date_ymd = parse_date_time(string_ymd, orders = "ymd"),
+        parsed_date_dmy = parse_date_time(string_dmy, orders = "dmy"),
+        parsed_date_mdy = parse_date_time(string_mdy, orders = "mdy")
+      ) %>%
+      collect(),
+    test_dates
+  )
+
+  # locale not working properly on windows
+  # TODO revisit once https://issues.apache.org/jira/browse/ARROW-16399 is done
+  skip_on_os("windows")
```

Comment:
I assumed we set the `locale` to `"C"` globally, don't we? https://github.com/apache/arrow/blob/d6ca3e2e9e995ff42df2465484ebf86de853a136/r/tests/testthat/test-dplyr-funcs-datetime.R#L31-L33

Those tests will simply return `NA` on Windows, resulting in a difference to `expected`.

Reply:
For example, see https://github.com/apache/arrow/runs/6209999159?check_suite_focus=true#step:20:41081 before the split of the tests data frame into locale-dependent vs locale-independent formats.

Reply:
> Even if we accept that dates with months written out in them will not work on windows, Looking at those failures: they are happening only with text months, so there actually are test cases that we should be asserting are ok in there.

Not sure what you mean by this. We are testing the months as numbers separately and those tests are run for all platforms (except those that do not have `RE2`). The reason for splitting was exactly to avoid the situation where we skip too many tests (and skipping just those that are problematic, i.e. `"%b"` and `"%B"` formats for months)

Reply:
> We are testing the months as numbers separately and those tests are run for all platforms (except those that do not have RE2)

Good to know! When I looked at these tests + those failures + where the skip is that wasn't clear to me. Let's address that. If we are going to skip, let's skip a targeted test that only contains the cases we care about and not intermingle tests cases that are tested elsewhere that do end up passing.

But most importantly: I'm still not clear _why_ these tests fail. The jira that is reference in the comment about is about when the locale is something other than C. But when these tests are run isn't the locale C? If it is the case that _even in the C locale_ on windows, we cannot `strptime` with `%b` on windows we should have a jira for that + reference that explicitly here.

Reply:
I created [ARROW-16443](https://issues.apache.org/jira/browse/ARROW-16443) to look into the failures to parse a string with month names on Windows. 

> If we are going to skip, let's skip a targeted test that only contains the cases we care about and not intermingle tests cases that are tested elsewhere

I think this is exactly what I was doing, wasn't I?

Reply:
Yes, sorry I was looking at the original failures that had values for both `test_dates` and `test_dates_locale` mixed together, but looking at this again you do test just `test_dates_locale` here which is good. Also 💯 on getting a Jira for this issue + linking it.

And honestly — to make this even more obvious — since you only use `test_dates` and `test_dates_locale` you might consider using those definitions directly as the `tbl` argument for `compare_dplyr_binding`. That way one doesn't have to scroll up to see what's in either of them.

Reply:
Done.